datafusion/expr/src/expr_fn.rs

```diff
@@ -993,7 +993,27 @@ pub fn create_udwf(
 pub fn call_fn(name: impl AsRef<str>, args: Vec<Expr>) -> Result<Expr> {
     match name.as_ref().parse::<BuiltinScalarFunction>() {
         Ok(fun) => Ok(Expr::ScalarFunction(ScalarFunction::new(fun, args))),
-        Err(e) => Err(e),
+        Err(_) => {
+            // Constructing a `ScalarUDF` with only name and stub impl/return_type...
+            // This unresolved UDF will be resolved during analyzing using registered functions
```

Comment:
If you can make this work, it seems like a good idea to me. 

However, it is not clear to me how the subsequent passes will know they have to resolve this particular scalar UDF though 🤔  so I am not sure this approach is viable

Specifically, how would the code look that the analysis pass used to test if a `ScalarUDF`  was unresolved or not?

Reply:
That's a great point, without enum the implementation seems a bit vulnerable to bugs
I think we can either make `Expr::ScalarFunction` or `Expr::ScalarUDF` a enum, since the long-term goal is to remove `BuiltinScalarFunction` they should be same.
I'll check which way can make future migration easier
```rust
pub enum ScalarFunctionDefinition {
  /// Resolved to a user defined function
  UDF(ScalarUDF),
  /// A scalar function that will be called by name
  Name(Arc<str>),
}

#[derive(Clone, PartialEq, Eq, Hash, Debug)]
pub struct ScalarUDF {
    /// The function
    pub fun: ScalarFunctionDefinition,
    /// List of expressions to feed to the functions as arguments
    pub args: Vec<Expr>,
}
```

```rust
pub enum ScalarFunctionDefinition {
  /// Resolved to a built in scalar function
  /// (will be removed long term)
  BuiltIn(built_in_function::BuiltinScalarFunction),
  /// Resolved to a user defined function
  UDF(ScalarUDF),
  /// A scalar function that will be called by name
  Name(Arc<str>),
}

#[derive(Clone, PartialEq, Eq, Hash, Debug)]
pub struct ScalarFunction {
    /// The function
    pub fun: ScalarFunctionDefinition,
    /// List of expressions to feed to the functions as arguments
    pub args: Vec<Expr>,
}
```